src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+
         switch ($command->getCartRuleType()) {
-            // Percent type
             case OrderDiscountType::DISCOUNT_PERCENT:
-                if ($discountValue < 100) {
-                    if (isset($orderInvoice)) {
-                        $cartRules[$orderInvoice->id]['value_tax_incl'] = Tools::ps_round(
-                            $orderInvoice->total_paid_tax_incl * $discountValue / 100,
-                            2
-                        );
-                        $cartRules[$orderInvoice->id]['value_tax_excl'] = Tools::ps_round(
-                            $orderInvoice->total_paid_tax_excl * $discountValue / 100,
-                            2
-                        );
-
-                        // Update OrderInvoice
-                        $this->applyDiscountOnInvoice(
-                            $orderInvoice,
-                            $cartRules[$orderInvoice->id]['value_tax_incl'],
-                            $cartRules[$orderInvoice->id]['value_tax_excl']
-                        );
-                    } elseif ($order->hasInvoice()) {
-                        $orderInvoices = $order->getInvoicesCollection();
-                        foreach ($orderInvoices as $orderInvoice) {
-                            /* @var OrderInvoice $orderInvoice */
-                            $cartRules[$orderInvoice->id]['value_tax_incl'] = Tools::ps_round(
-                                $orderInvoice->total_paid_tax_incl * $discountValue / 100,
-                                2
-                            );
-                            $cartRules[$orderInvoice->id]['value_tax_excl'] = Tools::ps_round(
-                                $orderInvoice->total_paid_tax_excl * $discountValue / 100,
-                                2
-                            );
-
-                            // Update OrderInvoice
-                            $this->applyDiscountOnInvoice(
-                                $orderInvoice,
-                                $cartRules[$orderInvoice->id]['value_tax_incl'],
-                                $cartRules[$orderInvoice->id]['value_tax_excl']
-                            );
-                        }
-                    } else {
-                        $cartRules[0]['value_tax_incl'] = Tools::ps_round(
-                            $order->total_paid_tax_incl * $discountValue / 100,
-                            2
-                        );
-                        $cartRules[0]['value_tax_excl'] = Tools::ps_round(
-                            $order->total_paid_tax_excl * $discountValue / 100,
-                            2
-                        );
-                    }
-                } else {
-                    throw new OrderException('Percentage discount value cannot be higher than 100%.');
-                }
+                $this->buildPercentTypeCartRules($cartRules, $order, $discountValue, $orderInvoice);
```

Comment:
I don't think you need to instanciate `$cartRules` beforehand and pass it around in each method Since you only deal with one cart rule type, only one method will be called, so each methods `buildPercentTypeCartRules`,  `buildAmountTypeCartRules` and `buildPercentTypeCartRules` can create its own internal array and return it.

This way (if you already made the simplification I suggested earlier with the array of invoices) you can simply do:

```php
switch ($cartRuleType) {
    case OrderDiscountType::DISCOUNT_PERCENT:
        return $this->buildPercentTypeCartRules($orderInvoices, $discountValue);
        break;
    ....
}
```